google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb

```diff
@@ -313,6 +314,351 @@ def data token: nil, max: nil, start: nil
         end
         alias query_results data
 
+        ##
+        # Yielded to a block to accumulate changes for a patch request.
+        class Updater < QueryJob
+          class << self
+            # If no job_id or prefix is given, always generate a client-side
+            # job ID anyway, for idempotent retry in the google-api-client
+            # layer. See
+            # https://cloud.google.com/bigquery/docs/managing-jobs#generate-jobid
+            def job_ref_from job_id, prefix
+              prefix ||= "job_"
+              job_id ||= "#{prefix}#{generate_id}"
+              API::JobReference.new(
+                project_id: @project,
+                job_id: job_id
+              )
+            end
+
+            def dataset_ref_from dts, pjt = nil
+              return nil if dts.nil?
+              if dts.respond_to? :dataset_id
+                Google::Apis::BigqueryV2::DatasetReference.new(
+                  project_id: (pjt || dts.project_id || @project),
+                  dataset_id: dts.dataset_id
+                )
+              else
+                Google::Apis::BigqueryV2::DatasetReference.new(
+                  project_id: (pjt || @project),
+                  dataset_id: dts
+                )
+              end
+            end
+          end
+
+          ##
```

Comment:
This should be marked `@private`.

Reply:
Done in https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/2003/commits/1729e02fa0ba50a3ea5442c8478e04328bae4cc3